vehicle/src/Vehicle/Compile/Type/System.hs

```diff
@@ -39,6 +39,13 @@ class (Eq builtin, Hashable builtin, NormalisableBuiltin builtin, TypableBuiltin
     Type builtin ->
     m (Type builtin)
 
+  restrictRecordType ::
+    (MonadTypeChecker builtin m) =>
+    DeclProvenance ->
+    Type builtin ->
+    [RecordField (Type builtin)] ->
+    m (Type builtin)
```

Comment:
Also we don't actually edit the type ever right so this could just return `m ()`